modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/aware/SegmentCompressStorage.java

```diff
@@ -79,10 +79,31 @@ synchronized void onSegmentCompressed(long compressedIdx) {
         else
             this.lastCompressedIdx = lastMaxCompressedIdx;
 
+        notifyAll();
+
         if (compressedIdx > lastEnqueuedToCompressIdx)
             lastEnqueuedToCompressIdx = compressedIdx;
     }
 
+    /**
+     * Method will wait activation of particular WAL segment index.
+     *
+     * @param awaitIdx absolute index {@link #lastCompressedIdx()}} to become true.
+     * @throws IgniteInterruptedCheckedException if interrupted.
+     */
+    public synchronized void awaitSegmentCompressed(long awaitIdx) throws IgniteInterruptedCheckedException {
+        while (lastCompressedIdx() < awaitIdx && !interrupted) {
+            try {
+                wait(2000);
```

Comment:
Is it possible to avoid blocking here? I suggest that this method should return a `future`, and calling thread just assign a listener on this future. 

This is especially important, because this blocking thread is part of the snapshotExecutorService pool, that is also used by `SnapshotPartitionsVerifyHandler`, it checks snapshot consistency. Then running snapshot operation might block independent task of checking existing snapshot.

I suppose at least we must create a ticket for that. WDYT?

Reply:
> I suppose at least we must create a ticket for that. WDYT?

https://issues.apache.org/jira/browse/IGNITE-17978